equinox/nn/_embedding.py

```diff
@@ -89,3 +96,313 @@ def __call__(
+                                         in_axes=1,
+                                         out_axes=1)(key_heads)
+
+                    return query_heads, key_heads, value_heads
+
+                mha_output = self.mha_attention(
+                    process_heads=process_heads,
+                    query=jax.vmap(self.rms_norm)(x),
+                    key_=jax.vmap(self.rms_norm)(x),
+                    value=jax.vmap(self.rms_norm)(x),
+                    mask=mask,
+                )
+        ```
+
+    ??? cite
+
+        [Attention is All You Need](https://arxiv.org/abs/1706.03762)
+
+        ```bibtex
+        @inproceedings{vaswani2017attention,
+            author={Vaswani, Ashish and Shazeer, Noam and Parmar, Niki and
+                    Uszkoreit, Jakob and Jones, Llion and Gomez, Aidan N and
+                    Kaiser, {\L}ukasz and Polosukhin, Illia},
+            booktitle={Advances in Neural Information Processing Systems},
+            publisher={Curran Associates, Inc.},
+            title={Attention is All You Need},
+            volume={30},
+            year={2017}
+        }
+        ```
+    """
+
+    embedding_size: int = field(static=True)
+    theta: float = field(static=True)
+
+    def __init__(
+        self,
+        embedding_size: int,
+        *,
+        theta: float = 10000.0,
+        key: Optional[PRNGKeyArray] = None,
+        **kwargs,
+    ):
+        """**Arguments:**
+        `SinusoidalPositionalEmbedding` requires:
+
+        - `embedding_size`: Size of the token embeddings. Must be non-negative.
+        - `theta`: The frequency of the sinusoidal positional encoding.
+            Must be positive. Defaults to 10000.0.
+        - `key`: Not used; provided for compatibility with the rest of the Equinox API.
+            (Keyword only argument.)
+        """
+
+        if embedding_size % 2 != 0:
+            raise ValueError(
+                f"embedding_size must be even, but {embedding_size} is not even."
+            )
+
+        if embedding_size <= 0:
+            raise ValueError(
+                f"embedding_size must be positive, but {embedding_size} <= 0."
+            )
+
+        self.embedding_size = embedding_size
+        self.theta = theta
+
+    @staticmethod
+    def get_positional_encoding(
+        embedding_size: int, seq_len: int, theta: float = 10000.0
+    ) -> Float[Array, "seq_len embedding_size"]:
+        pos = jnp.arange(seq_len)[:, jnp.newaxis]
+        div_term = jnp.exp(
+            jnp.arange(0, embedding_size, 2) * -(jnp.log(theta) / embedding_size)
+        )
+        # the following expression is closer to the actual notation they used.
+        # div_term = 1 / 10000 ** (jnp.arange(0, embedding_size, 2) / embedding_size)
+        pos_enc = jnp.zeros((seq_len, embedding_size))
+        pos_enc = pos_enc.at[:, 0::2].set(jnp.sin(pos * div_term))
+        pos_enc = pos_enc.at[:, 1::2].set(jnp.cos(pos * div_term))
```

Comment:
I have a pretty good hunch that JAX is going to do a terrible job of interleaving these, and will probably make copies here.

It doesn't *really* matter since we're running this under `jax.ensure_compile_time_eval`, but if I had to guess, then probably
```python
mult = pos * div_term
sines = jnp.sin(mult)
cosines = jnp.cos(mult)
jnp.stack([sines, cosines], axis=2).reshape((seq_len, embedding_size))
```
will be much faster.
(Assuming what I've written about is functionally correct anyway, please double-check...)

By the way, I'd have thought the `pos * div_term` broadcasting should fail when running under `JAX_NUMPY_RANK_PROMOTION=raise`, as [we do](https://github.com/patrick-kidger/equinox/blob/34252989c251c315c6eee5895dd64e9645b350ce/tests/conftest.py#L11). I'm surprised it doesn't seem to here - any ideas?

Reply:
Regarding your code: it's indeed 2x faster (unjitted) and _a bit_ faster when jitted (outputs are exactly the same). 

Mine: 
```
181 µs ± 945 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
```

Yours:
```
150 µs ± 569 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
```

Since I don't see any downsides here, I'll change it to your implementation.

As for the rank promotions: When I added those into jupyterlab to test everything, it actually does raise issues! But it does that for the RoPE code too. I'll fix those of course but I'm not sure why it's not raised by the GH test runner.